messaging/broker.go

```diff
@@ -144,943 +222,944 @@ func (b *Broker) closeTopics() {
-
-		// Append replica's topics.
-		for _, srt := range sr.Topics {
-			r.topics[srt.TopicID] = srt.Index
-		}
-	}
 
-	// Read the highest index from each of the topic files.
-	if err := b.loadIndex(); err != nil {
-		return fmt.Errorf("load index: %s", err)
-	}
+	// Set in-memory index.
+	b.index = index
 
 	return nil
 }
 
-// loadIndex reads through all topics to find the highest known index.
-func (b *Broker) loadIndex() error {
-	for _, t := range b.topics {
-		if err := t.loadIndex(); err != nil {
-			return fmt.Errorf("topic(%d): %s", t.id, err)
-		} else if t.index > b.index {
-			b.index = t.index
-		}
-	}
-	return nil
-}
-
-// save persists the broker metadata to disk.
-func (b *Broker) save() error {
-	if b.path == "" {
-		return ErrClosed
-	}
+// Snapshot streams the current state of the broker and returns the index.
+func (b *Broker) Snapshot(w io.Writer) (uint64, error) {
+	// TODO: Prevent truncation during snapshot.
 
 	// Calculate header under lock.
+	b.mu.RLock()
 	hdr, err := b.createSnapshotHeader()
+	b.mu.RUnlock()
 	if err != nil {
-		return fmt.Errorf("create snapshot: %s", err)
+		return 0, fmt.Errorf("create snapshot: %s", err)
 	}
 
-	// Write snapshot to disk.
-	f, err := os.Create(b.metaPath())
+	// Encode snapshot header.
+	buf, err := json.Marshal(&hdr)
 	if err != nil {
-		return err
+		return 0, fmt.Errorf("encode snapshot header: %s", err)
 	}
-	defer func() { _ = f.Close() }()
 
-	// Write snapshot to disk.
-	if err := json.NewEncoder(f).Encode(&hdr); err != nil {
-		return err
+	// Write header frame.
+	if err := binary.Write(w, binary.BigEndian, uint32(len(buf))); err != nil {
+		return 0, fmt.Errorf("write header size: %s", err)
+	}
+	if _, err := w.Write(buf); err != nil {
+		return 0, fmt.Errorf("write header: %s", err)
 	}
 
-	return nil
-}
-
-// mustSave persists the broker metadata to disk. Panic on error.
-func (b *Broker) mustSave() {
-	if err := b.save(); err != nil && err != ErrClosed {
-		panic(err.Error())
+	// Stream each topic sequentially.
+	for _, t := range hdr.Topics {
+		for _, s := range t.Segments {
+			if _, err := copyFileN(w, s.path, s.Size); err != nil {
+				return 0, err
+			}
+		}
 	}
+
+	// Return the snapshot and its last applied index.
+	return hdr.Index, nil
 }
 
 // createSnapshotHeader creates a snapshot header.
 func (b *Broker) createSnapshotHeader() (*snapshotHeader, error) {
 	// Create parent header.
-	s := &snapshotHeader{}
+	sh := &snapshotHeader{Index: b.index}
 
 	// Append topics.
 	for _, t := range b.topics {
-		// Retrieve current topic file size.
-		var sz int64
-		if t.file != nil {
-			fi, err := t.file.Stat()
-			if err != nil {
-				return nil, err
-			}
-			sz = fi.Size()
-		}
+		// Create snapshot topic.
+		st := &snapshotTopic{ID: t.id}
 
-		// Append topic to the snapshot.
-		s.Topics = append(s.Topics, &snapshotTopic{
-			ID:    t.id,
-			Index: t.index,
-			Size:  sz,
-			path:  t.path,
-		})
-	}
+		// Read segments from disk, not topic.
```

Comment:
Minor: I don't get this comment. Where else could segments come from? They are on the disk, and make up topics. I don't think there is any confusion, or am I missing something?